### ISSUE TITLE
Problems removing dead nodes with a missing data center.

### ISSUE THREAD

**Issue by fridgei** [scylladb/scylladb#11078]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 4.6.3
Cluster size: 60
OS (RHEL/CentOS/Ubuntu/AWS AMI): `AWS AMI 4.14.243-185.433.amzn2.x86_64`

## Issue Description
We created a new data center in our scylla cluster and migrated data from our legacy cluster us-east to the new cluster us-east-v1. After the migration was completed a spurious update to the managed node group occurred and all of the ec2s were replaced rendering all 15 nodes in the new us-east-v1 to be permanently down and unavailable.  Initially I tried to do one by one node replacements by setting `replace_address_first_boot`.  These operations all failed and the new nodes would not join the cluster.  I then tried to use `nodetool removenode` and was unable to perform the operations.  I initially tried to do this with the cli and did not list the dead nodes.

```root@ip-192-168-203-231:/# nodetool removenode 4bd4ec5f-a121-40d1-8c2b-6d1eeb8a57c6
nodetool: Scylla API server HTTP POST to URL '/storage_service/remove_node' failed: std::runtime_error (node_ops_cmd_check: Node 192.168.211.130 rejected node_ops_cmd=3 from node=192.168.203.231 with ops_uuid=ba5099e3-5335-48e6-b28e-d6f2c295877d, pending_node_ops={}, the node ops is unknown)```

Then we tried again ignoring the down nodes 
```root@ip-192-168-203-231:/# nodetool removenode --ignore-dead-nodes 192.168.151.141,192.168.182.130,192.168.167.197,192.168.201.184,192.168.163.92,192.168.202.80,192.168.254.48,192.168.246.177,192.168.188.114,192.168.131.147,192.168.198.147,192.168.144.116,192.168.247.181  4bd4ec5f-a121-40d1-8c2b-6d1eeb8a57c6
nodetool: Scylla API server HTTP POST to URL '/storage_service/remove_node' failed: std::runtime_error (node_ops_cmd_check: Node 192.168.211.130 rejected node_ops_cmd=3 from node=192.168.203.231 with ops_uuid=240b0ef6-ee8d-4524-ab39-8f71df86fccf, pending_node_ops={}, the node ops is unknown)```
We then tried to perform the removal by hitting the rest endpoint also and this did not work.

**Comment by fee-mendes** [comment]
@asias `node_ops_cmd=3` should be `removenode_abort` isn't it? Why would it fail even when specifying `--ignore-dead-nodes` ?

**Comment by asias** [comment]
Hello @fridgei, we saw similar issue  https://github.com/scylladb/scylla/issues/10337. 

I have sent PRs for this issue. Some of  them  are already merged. Some are on the way to master branch.

I plan to backport them to 4.6 branch.

**Comment by asias** [comment]
> @asias `node_ops_cmd=3` should be `removenode_abort` isn't it? Why would it fail even when specifying `--ignore-dead-nodes` ?

The procedure was:

Node 192.168.211.130 did not get the heart beat update from 192.168.203.231 so it removed the node ops with uuid 240b0ef6-ee8d-4524-ab39-8f71df86fccf.

Node 192.168.203.231 sent removenode_abort to all live nodes including 192.168.211.130 to revert the cluster to a state before removenode ops when 192.168.203.231 failed the remove node ops.

Node 192.168.211.130 reported it did not know ops uuid 240b0ef6-ee8d-4524-ab39-8f71df86fccf any more.

**Comment by asias** [comment]
In addition to:

https://github.com/scylladb/scylla/pull/10850
https://github.com/scylladb/scylla/pull/10822


I just sent another PR:

https://github.com/scylladb/scylla/pull/11081

**Comment by slivne** [comment]
@asias
- anything left here or can we close this ?
- what backports are needed to 5.1/5.0/4.6 ?

**Comment by asias** [comment]
We need to backport:

From Asias:
```
https://github.com/scylladb/scylladb/pull/10850
https://github.com/scylladb/scylladb/pull/10822
https://github.com/scylladb/scylladb/pull/11081
``

From Benny:
```
commit 9167b857e9c3f32032b6e9a6eed73610c0da0a85
Author: Benny Halevy <bhalevy@scylladb.com>
Date:   Wed Aug 10 23:18:46 2022 +0300

    abstract_replication_strategy: calculate_effective_replication_map: optimize for static replication strategies
    
    For replication strategies like "everywhere"
    and "local" that return the same set of endpoints
    for all tokens, we can call rs->calculate_natural_endpoints
    one once and reuse the result for all token.
    
    Note that ideally the replication_map could contain only
    a single token range for this case, but that does't seem to work yet.
    
    Add maybe_yield() calls to the tight loop
    to prevent reactor stalls on large clusters when copying
    a long vector returned by everywhere_replication_strategy
    to potentially 1000's of tokens in the map.
    
    Nicholas Peshek wrote in
    https://github.com/scylladb/scylladb/issues/10337#issuecomment-1211152370
    
    about similar patch by Geoffrey Beausire:
    https://github.com/criteo-forks/scylla/commit/994c6ecf3c4edd151cd1db2ed0c92f32dc832cbb
    
    > Yep. That dropped our startup from 3000+ seconds to about 40.
    
    Fixes #10337
    
    Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

commit eb678e723b0056554dcfc49d605b732e45eb19ab
Author: Benny Halevy <bhalevy@scylladb.com>
Date:   Wed Aug 10 23:09:49 2022 +0300

    abstract_replication_strategy: add has_uniform_natural_endpoints
    
    So that using calaculate_natural_endpoints can be optimized
    for strategies that return the same endpoints for all tokens,
    namely everywhere_replication_strategy and local_strategy.
    
    Signed-off-by: Benny Halevy <bhalevy@scylladb.com>
```

**Comment by asias** [comment]
I am closing this one since the later releases contain all the fixes mentioned above. Please reopen if you see the problems again. Thanks.
